File: android/DebugSocket.kt
// SPDX-License-Identifier: MIT
// Copyright © 2026 doxx.net. All Rights Reserved.

package net.doxx.debugsocket

import android.annotation.SuppressLint
import android.os.Build
import android.provider.Settings
import android.util.Base64
import android.util.Log
import kotlinx.coroutines.*
import okhttp3.*
import okio.ByteString
import org.json.JSONObject
import java.security.MessageDigest
import java.security.SecureRandom
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit
import javax.net.ssl.*

/**
 * Remote debug log streaming for debug/beta builds.
 * Drop this file into your Android project and call DebugSocket.connectIfDebug(context) on app launch.
 *
 * Dependencies (add to build.gradle):
 *   implementation 'com.squareup.okhttp3:okhttp:4.12.0'
 *   implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-android:1.7.3'
 */
object DebugSocket {
    private const val TAG = "DebugSocket"

    // ========================================
    // CONFIGURATION - UPDATE THESE
    // ========================================
    private const val SERVER_URL = "wss://debug.doxx.net/stream"
    private const val SHARED_SECRET = "YOUR_SHARED_SECRET_HERE"

    // Certificate pinning (optional)
    // Set to null to use standard PKI validation
    // Set to base64-encoded DER certificate to pin to a specific cert
    // Generate with: make gencert (on the server side)
    private val PINNED_CERTIFICATE_BASE64: String? = null

    // Example pinned cert:
    // private val PINNED_CERTIFICATE_BASE64: String? = """
    // MIIB4zCCAYigAwIBAgIRAPAiLKkxJU39GLFOMYWqQNMwCgYIKoZIzj0EAwIwODEd
    // ... (paste from gencert output) ...
    // """.trimIndent()

    // ========================================
    // State
    // ========================================
    private var webSocket: WebSocket? = null
    private var client: OkHttpClient? = null
    private var isConnected = false
    private var shouldReconnect = true
    private var deviceHash: String? = null
    private var deviceName: String? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val isoDateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }

    // ========================================
    // Public API
    // ========================================

    /**
     * Call this on app launch - only connects for debug builds
     * @param context Android context (Application or Activity)
     */
    @SuppressLint("HardwareIds")
    fun connectIfDebug(context: android.content.Context) {
        // Only connect in debug builds
        if (!BuildConfig.DEBUG) {
            // For release builds, you could check if it's a beta/internal track
            // For now, skip entirely in release
            return
        }

        // Get device identifier
        val androidId = Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID)
        deviceHash = md5Hash(androidId ?: UUID.randomUUID().toString())
        deviceName = "${Build.MANUFACTURER} ${Build.MODEL}"

        connect()
    }

    /**
     * Manually start connection
     */
    fun connect() {
        if (deviceHash == null) {
            Log.e(TAG, "Must call connectIfDebug(context) first")
            return
        }

        scope.launch {
            doConnect()
        }
    }

    /**
     * Stop streaming and disconnect
     */
    fun disconnect() {
        shouldReconnect = false
        webSocket?.close(1000, "Client disconnect")
        webSocket = null
        isConnected = false
    }

    /**
     * Send a log message to the debug server
     */
    fun log(message: String) {
        if (!isConnected || webSocket == null) return

        try {
            val entry = JSONObject().apply {
                put("ts", isoDateFormat.format(Date()))
                put("msg", message)
            }
            webSocket?.send(entry.toString())
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send log: ${e.message}")
        }
    }

    /**
     * Log with tag prefix (convenience method)
     */
    fun log(tag: String, message: String) {
        log("[$tag] $message")
    }

    // ========================================
    // Private Implementation
    // ========================================

    private fun doConnect() {
        if (webSocket != null) return

        val hash = deviceHash ?: return
        val name = deviceName?.let { java.net.URLEncoder.encode(it, "UTF-8") } ?: "Unknown"

        var urlString = "$SERVER_URL?device=$hash&name=$name&secret=$SHARED_SECRET"

        // Add tunnel IPs if available (hook into your VPN manager)
        getCurrentIPv4()?.let { urlString += "&ipv4=$it" }
        getCurrentIPv6()?.let { urlString += "&ipv6=$it" }

        // Build OkHttp client
        client = buildOkHttpClient()

        val request = Request.Builder()
            .url(urlString)
            .build()

        webSocket = client?.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                isConnected = true
                shouldReconnect = true
                Log.i(TAG, "Connected to $SERVER_URL")
                if (PINNED_CERTIFICATE_BASE64 != null) {
                    Log.i(TAG, "Using pinned certificate")
                }
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                // Server might send commands in future
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                // Binary messages not used
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                Log.i(TAG, "Connection closing: $code $reason")
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                handleDisconnect()
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e(TAG, "Connection failed: ${t.message}")
                handleDisconnect()
            }
        })
    }

    private fun handleDisconnect() {
        isConnected = false
        webSocket = null

        if (!shouldReconnect) return

        Log.i(TAG, "Disconnected, reconnecting in 5s...")
        scope.launch {
            delay(5000)
            doConnect()
        }
    }

    private fun buildOkHttpClient(): OkHttpClient {
        val builder = OkHttpClient.Builder()
            .pingInterval(30, TimeUnit.SECONDS)
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(0, TimeUnit.SECONDS) // No timeout for WebSocket
            .writeTimeout(30, TimeUnit.SECONDS)

        // Apply certificate pinning if configured
        PINNED_CERTIFICATE_BASE64?.let { base64Cert ->
            try {
                val certBytes = Base64.decode(base64Cert.replace("\n", "").replace(" ", ""), Base64.DEFAULT)
                val certFactory = CertificateFactory.getInstance("X.509")
                val certificate = certFactory.generateCertificate(certBytes.inputStream()) as X509Certificate

                // Create trust manager that only trusts the pinned cert
                val trustManager = object : X509TrustManager {
                    override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?) {}

                    override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?) {
                        if (chain.isNullOrEmpty()) {
                            throw javax.net.ssl.SSLException("No server certificate")
                        }
                        // Compare the server's certificate with our pinned certificate
                        if (!chain[0].encoded.contentEquals(certificate.encoded)) {
                            throw javax.net.ssl.SSLException("Certificate pinning: MISMATCH")
                        }
                        Log.d(TAG, "Certificate pinning: MATCHED")
                    }

                    override fun getAcceptedIssuers(): Array<X509Certificate> = arrayOf(certificate)
                }

                val sslContext = SSLContext.getInstance("TLS")
                sslContext.init(null, arrayOf(trustManager), SecureRandom())

                builder.sslSocketFactory(sslContext.socketFactory, trustManager)
                builder.hostnameVerifier { _, _ -> true } // We're pinning, so hostname doesn't matter

            } catch (e: Exception) {
                Log.e(TAG, "Failed to setup certificate pinning: ${e.message}")
            }
        }

        return builder.build()
    }

    private fun md5Hash(input: String): String {
        val md = MessageDigest.getInstance("MD5")
        val digest = md.digest(input.toByteArray())
        return digest.joinToString("") { "%02x".format(it) }
    }

    /**
     * Override this to provide current tunnel IPv4 address
     */
    private fun getCurrentIPv4(): String? {
        // TODO: Hook into your VPN/tunnel manager
        // Example: return VpnManager.currentTunnel?.assignedIP
        return null
    }

    /**
     * Override this to provide current tunnel IPv6 address
     */
    private fun getCurrentIPv6(): String? {
        // TODO: Hook into your VPN/tunnel manager
        // Example: return VpnManager.currentTunnel?.assignedIPv6
        return null
    }
}

// ========================================
// Integration Example
// ========================================

/*
 * 1. Add dependencies to build.gradle:
 *
 *    dependencies {
 *        implementation 'com.squareup.okhttp3:okhttp:4.12.0'
 *        implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-android:1.7.3'
 *    }
 *
 * 2. In your Application class:
 *
 *    class MyApp : Application() {
 *        override fun onCreate() {
 *            super.onCreate()
 *            DebugSocket.connectIfDebug(this)
 *        }
 *    }
 *
 * 3. Use in your code:
 *
 *    // Simple log
 *    DebugSocket.log("User tapped connect button")
 *
 *    // Log with tag
 *    DebugSocket.log("VPN", "Tunnel established: $serverName")
 *
 *    // Wrap existing logging
 *    fun myLog(tag: String, msg: String) {
 *        Log.d(tag, msg)
 *        DebugSocket.log(tag, msg)
 *    }
 */

File: swift/DebugSocket.swift
// SPDX-License-Identifier: MIT
// Copyright © 2026 doxx.net. All Rights Reserved.

import Foundation
import UIKit
import Security

/// Remote debug log streaming for TestFlight builds
/// Drop this file into your iOS project and call DebugSocket.shared.connectIfTestFlight() on app launch
public class DebugSocket: NSObject {
    public static let shared = DebugSocket()

    private var socket: URLSessionWebSocketTask?
    private var session: URLSession?
    private let queue = DispatchQueue(label: "com.doxx.debugsocket", qos: .utility)
    private var isConnected = false
    private var shouldReconnect = true

    // MARK: - Configuration (UPDATE THESE)
    
    private let serverURL = "wss://debug.doxx.net/stream"  // Your server URL
    private let sharedSecret = "YOUR_SHARED_SECRET_HERE"   // Must match --secret on server
    
    // Certificate pinning (optional)
    // Set to nil to use standard PKI validation
    // Set to base64-encoded DER certificate to pin to a specific cert
    private let pinnedCertificateBase64: String? = nil
    
    // Example pinned cert (replace with your own from: make gencert):
    // private let pinnedCertificateBase64: String? = """
    // MIIBkTCB+wIJAKHBfpegPfHtMA0GCSqGSIb3DQEBCwUAMBExDzANBgNVBAMMBmRl
    // YnVnczAeFw0yNjAxMjYwMDAwMDBaFw0zNjAxMjYwMDAwMDBaMBExDzANBgNVBAMM
    // BmRlYnVnczBcMA0GCSqGSIb3DQEBAQUAA0sAMEgCQQC7o... (truncated)
    // """
    
    private override init() {
        super.init()
    }

    // MARK: - Public API

    /// Call this on app launch - only connects for TestFlight/Debug builds
    public func connectIfTestFlight() {
        #if DEBUG
        connect()
        #else
        // Check if this is a TestFlight build (sandbox receipt)
        if Bundle.main.appStoreReceiptURL?.lastPathComponent == "sandboxReceipt" {
            connect()
        }
        #endif
    }

    /// Manually start connection
    public func connect() {
        queue.async { [weak self] in
            self?.doConnect()
        }
    }

    /// Stop streaming and disconnect
    public func disconnect() {
        shouldReconnect = false
        socket?.cancel(with: .goingAway, reason: nil)
        socket = nil
        isConnected = false
    }

    /// Send a log message to the debug server
    public func log(_ message: String) {
        guard isConnected, let socket = socket else { return }

        let entry: [String: Any] = [
            "ts": ISO8601DateFormatter().string(from: Date()),
            "msg": message
        ]

        guard let data = try? JSONSerialization.data(withJSONObject: entry),
              let json = String(data: data, encoding: .utf8) else { return }

        socket.send(.string(json)) { error in
            if error != nil {
                // Connection might be dead, will reconnect
            }
        }
    }

    // MARK: - Private

    private func doConnect() {
        guard socket == nil else { return }

        let deviceHash = DeviceIdentification.deviceHash
        let deviceName = UIDevice.current.name
            .addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? "Unknown"

        var urlString = "\(serverURL)?device=\(deviceHash)&name=\(deviceName)&secret=\(sharedSecret)"

        // Add tunnel IPs if available (hook into your TunnelManager)
        if let ipv4 = getCurrentIPv4() {
            urlString += "&ipv4=\(ipv4)"
        }
        if let ipv6 = getCurrentIPv6() {
            urlString += "&ipv6=\(ipv6)"
        }

        guard let url = URL(string: urlString) else {
            print("[DebugSocket] Invalid URL")
            return
        }

        // Create session with optional certificate pinning
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 300
        
        if pinnedCertificateBase64 != nil {
            // Use delegate for certificate pinning
            session = URLSession(configuration: config, delegate: self, delegateQueue: nil)
        } else {
            // Standard PKI validation
            session = URLSession(configuration: config)
        }
        
        socket = session?.webSocketTask(with: url)
        socket?.resume()
        isConnected = true
        shouldReconnect = true

        print("[DebugSocket] Connected to \(serverURL)")
        if pinnedCertificateBase64 != nil {
            print("[DebugSocket] Using pinned certificate")
        }

        // Keep connection alive with ping
        schedulePing()
        receiveLoop()
    }

    /// Override this to provide the current tunnel IPv4 address
    private func getCurrentIPv4() -> String? {
        // TODO: Hook into your TunnelManager to get assigned IPv4
        // Example: return TunnelManager.shared.currentTunnel?.assignedIP
        return nil
    }

    /// Override this to provide the current tunnel IPv6 address
    private func getCurrentIPv6() -> String? {
        // TODO: Hook into your TunnelManager to get assigned IPv6
        // Example: return TunnelManager.shared.currentTunnel?.assignedIPv6
        return nil
    }

    private func schedulePing() {
        queue.asyncAfter(deadline: .now() + 30) { [weak self] in
            guard let self = self, self.isConnected else { return }
            self.socket?.sendPing { error in
                if error == nil {
                    self.schedulePing()
                } else {
                    self.handleDisconnect()
                }
            }
        }
    }

    private func receiveLoop() {
        socket?.receive { [weak self] result in
            guard let self = self else { return }
            switch result {
            case .success:
                // Keep listening (server might send commands in future)
                self.receiveLoop()
            case .failure:
                self.handleDisconnect()
            }
        }
    }

    private func handleDisconnect() {
        isConnected = false
        socket?.cancel(with: .goingAway, reason: nil)
        socket = nil

        guard shouldReconnect else { return }

        // Retry after 5 seconds
        print("[DebugSocket] Disconnected, reconnecting in 5s...")
        queue.asyncAfter(deadline: .now() + 5) { [weak self] in
            self?.doConnect()
        }
    }
    
    // MARK: - Certificate Pinning Helper
    
    private func loadPinnedCertificate() -> SecCertificate? {
        guard let base64 = pinnedCertificateBase64,
              let data = Data(base64Encoded: base64.replacingOccurrences(of: "\n", with: "")
                                                    .replacingOccurrences(of: " ", with: "")) else {
            return nil
        }
        return SecCertificateCreateWithData(nil, data as CFData)
    }
}

// MARK: - URLSessionDelegate for Certificate Pinning

extension DebugSocket: URLSessionDelegate {
    public func urlSession(_ session: URLSession,
                          didReceive challenge: URLAuthenticationChallenge,
                          completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        
        guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
              let serverTrust = challenge.protectionSpace.serverTrust else {
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }
        
        // If no pinned certificate, use default validation
        guard let pinnedCert = loadPinnedCertificate() else {
            completionHandler(.performDefaultHandling, nil)
            return
        }
        
        // Get server certificate
        let serverCertCount = SecTrustGetCertificateCount(serverTrust)
        guard serverCertCount > 0 else {
            print("[DebugSocket] No server certificate found")
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }
        
        // Compare certificates
        if #available(iOS 15.0, *) {
            // iOS 15+ uses SecTrustCopyCertificateChain
            guard let certChain = SecTrustCopyCertificateChain(serverTrust) as? [SecCertificate],
                  let serverCert = certChain.first else {
                completionHandler(.cancelAuthenticationChallenge, nil)
                return
            }
            
            let serverCertData = SecCertificateCopyData(serverCert) as Data
            let pinnedCertData = SecCertificateCopyData(pinnedCert) as Data
            
            if serverCertData == pinnedCertData {
                print("[DebugSocket] Certificate pinning: MATCHED")
                completionHandler(.useCredential, URLCredential(trust: serverTrust))
            } else {
                print("[DebugSocket] Certificate pinning: MISMATCH - rejecting connection")
                completionHandler(.cancelAuthenticationChallenge, nil)
            }
        } else {
            // iOS 14 and earlier (deprecated API)
            guard let serverCert = SecTrustGetCertificateAtIndex(serverTrust, 0) else {
                completionHandler(.cancelAuthenticationChallenge, nil)
                return
            }
            
            let serverCertData = SecCertificateCopyData(serverCert) as Data
            let pinnedCertData = SecCertificateCopyData(pinnedCert) as Data
            
            if serverCertData == pinnedCertData {
                print("[DebugSocket] Certificate pinning: MATCHED")
                completionHandler(.useCredential, URLCredential(trust: serverTrust))
            } else {
                print("[DebugSocket] Certificate pinning: MISMATCH - rejecting connection")
                completionHandler(.cancelAuthenticationChallenge, nil)
            }
        }
    }
}

// MARK: - Integration with existing Logger

/*
 Add this to your wg_log() function in Logger.swift:

 func wg_log(_ type: OSLogType, message msg: String) {
     os_log("%{public}s", log: OSLog.default, type: type, msg)
     guard Logger.isEnabled else { return }
     Logger.global?.log(message: msg)

     // Stream to DebugSocket (add this line)
     DebugSocket.shared.log(msg)
 }

 And in your App.swift or AppDelegate:

 @main
 struct MyApp: App {
     init() {
         DebugSocket.shared.connectIfTestFlight()
     }
     // ...
 }
*/
